include/mbedtls/sm3.h

```diff
@@ -0,0 +1,180 @@
+ */
+void mbedtls_sm3_init( mbedtls_sm3_context *ctx );
+
+/**
+ * \brief          This function clears a SM3 context.
+ *
+ * \param ctx      The SM3 context to clear. This may be \c NULL, in which
+ *                 case this function returns immediately. If it is not \c NULL,
+ *                 it must point to an initialized SM3 context.
+ */
+void mbedtls_sm3_free( mbedtls_sm3_context *ctx );
+
+/**
+ * \brief          This function clones the state of a SM3 context.
+ *
+ * \param dst      The destination context. This must be initialized.
+ * \param src      The context to clone. This must be initialized.
+ */
+void mbedtls_sm3_clone( mbedtls_sm3_context *dst,
+                        const mbedtls_sm3_context *src );
+
+/**
+ * \brief          This function starts a SM3 checksum calculation.
+ *
+ * \param ctx      The context to use. This must be initialized.
+ *
+ * \return         \c 0 on success.
+ * \return         A negative error code on failure.
+ */
+int mbedtls_sm3_starts_ret( mbedtls_sm3_context *ctx );
+
+/**
+ * \brief          This function feeds an input buffer into an ongoing
+ *                 SM3 checksum calculation.
+ *
+ * \param ctx      The SM3 context. This must be initialized
+ *                 and have a hash operation started.
+ * \param input    The buffer holding the data. This must be a readable
+ *                 buffer of length \p ilen Bytes.
+ * \param ilen     The length of the input data in Bytes.
+ *
+ * \return         \c 0 on success.
+ * \return         A negative error code on failure.
+ */
+int mbedtls_sm3_update_ret( mbedtls_sm3_context *ctx,
+                            const unsigned char *input,
+                            size_t ilen );
+
+/**
+ * \brief          This function finishes the SM3 operation, and writes
+ *                 the result to the output buffer.
+ *
+ * \param ctx      The SM3 context. This must be initialized
+ *                 and have a hash operation started.
+ * \param output   The SM3 checksum result.
+ *                 This must be a writable buffer of length \c 32 Bytes.
+ *
+ * \return         \c 0 on success.
+ * \return         A negative error code on failure.
+ */
+int mbedtls_sm3_finish_ret( mbedtls_sm3_context *ctx,
+                            unsigned char output[32] );
+
+/**
+ * \brief          This function processes a single data block within
+ *                 the ongoing SM3 computation. This function is for
+ *                 internal use only.
+ *
+ * \param ctx      The SM3 context. This must be initialized.
+ * \param data     The buffer holding one block of data. This must
+ *                 be a readable buffer of length \c 64 Bytes.
+ *
+ * \return         \c 0 on success.
+ * \return         A negative error code on failure.
+ */
+int mbedtls_internal_sm3_process( mbedtls_sm3_context *ctx,
+                                  const unsigned char data[64] );
```

Comment:
```suggestion
/**
 * \brief          This function processes a single data block within
 *                 the ongoing SM3 computation. This function is for
 *                 internal use only.
 *
 * \param ctx      The SM3 context. This must be initialized.
 * \param data     The buffer holding one block of data. This must
 *                 be a readable buffer of length \c 64 Bytes.
 *
 * \return         \c 0 on success.
 * \return         A negative error code on failure.
 */
int mbedtls_internal_sm3_process( mbedtls_sm3_context *ctx,
                                  const unsigned char data[MBEDTLS_MD_MAX_SIZE] );
```

Reply:
> 
Can we keep

Reply:
> 
In ./include/mbedtls/md.h   MBEDTLS_MD_MAX_SIZE is *not* always 64:
      #if defined(MBEDTLS_SHA512_C)
      #define MBEDTLS_MD_MAX_SIZE         64  /* longest known is SHA512 */
      #else
      #define MBEDTLS_MD_MAX_SIZE         32  /* longest known is SHA256 or less */
      #endif

But SM3 block size should always be 64-bytes.

So may I suggest we keep a fixed 64-byte data buffer in?:
int mbedtls_internal_sm3_process( mbedtls_sm3_context *ctx, const unsigned char data[64] );
instead of:
int mbedtls_internal_sm3_process( mbedtls_sm3_context *ctx, const unsigned char data[MBEDTLS_MD_MAX_SIZE] );

Unless you are aware it is breaking code some checker rules?

Reply:
change MBEDTLS_MD_MAX_SIZE to
```
#if defined(MBEDTLS_SHA512_C) || defined(MBEDTLS_SM3_C)
#define MBEDTLS_MD_MAX_SIZE 64 /* longest known is SHA512 /
#else
#define MBEDTLS_MD_MAX_SIZE 32 / longest known is SHA256 or less */
#endif
```

Reply:
You are right. It is block size not DIGEST size.